Setting default detection parameters
----------------------------------------------------

##### Graph A/ Coverage impact on detection.

Increasing the coverage allows to quickly lose false candidates with a very limited impact on the number of true candidates (Arabidopsis thaliana data, powerlow abundance).

data/Graphics/PW_cat_depth_False_candidats_bySample.csv is the concatenation of 10 tests files (one by sample) depth_False_candidats.csv generated by data/Graphics/complete_candidat_file.sh .

Script to generate this graphic:

```
#!/usr/bin/env python3
import os
import subprocess
import pandas as pd
import re
import statsmodels
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

df_plot = pd.read_csv('data/PW_cat_depth_False_candidats_bySample.csv', sep=',')

layout = go.Layout() 
y_false=df_plot['False']
x_false=df_plot.index
	
fig=go.Figure(layout=layout) 
fig.add_trace(go.Box(x=df_plot["Cov"], y=df_plot["False"],  fillcolor ='rgb(241,90,96)', line_color='rgb(241,90,96)', name='Nb false'))
                    
fig.update_layout(boxmode='group', showlegend=True, paper_bgcolor='rgba(0,0,0,0)',
    plot_bgcolor='rgba(0,0,0,0)', font=dict(
        family="sans-serif",
        size=14), legend_font_size=14, xaxis=dict(
        title="Coverage", range=(0, 10)
    ),
    yaxis=dict(
        title="Nb false candidates"
    ))

# Legend position
fig.update_layout(legend=dict(
    orientation="v",
    yanchor="bottom",
    y=1.02,
    xanchor="right",
    x=1
    ))
    
fig.update_layout( yaxis=dict(
        autorange=True,
        showgrid=True,
        zeroline=True,
        dtick=5,
        gridcolor='rgb(211,211,211)',
        gridwidth=1,
        zerolinecolor='rgb(211,211,211)',
        zerolinewidth=2,
    ),
     paper_bgcolor='rgb(255,255,255)',
     plot_bgcolor='rgb(255,255,255)')

# Save plot     
if not os.path.exists("images"):
    os.mkdir("images")
fig.write_image("images/nb_false_by_coverage_PW_10samples.svg")
fig.write_html("images/nb_false_by_coverage_PW_10samples.html")
```

##### Graph B/ Filter candidates on border improve detection.

DPratio = DPin/(DPbefore+DPafter)

Input file 'data/Graphics/cat_pw_candidates.Cglen.intron.Cdlen.DPratio.RI.JC.csv' is used for B, C and D graphs.
Input file 'cat_pw_candidates.Cglen.intron.Cdlen.DPratio.RI.JC.csv' is the concatenation of 10 files (1 par sample) candidates.Cglen.intron.Cdlen.DPratio.RI.JC.csv, generated by data/Graphics/complete_candidat_file.sh .

Script to generate this graphic:

```
#!/usr/bin/env python3

import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import os

sns.set_theme(style="whitegrid")
df = pd.read_csv('data/cat_pw_candidates.Cglen.intron.Cdlen.DPratio.RI.JC.csv', sep=',')

# Colors list
colors_list=['#F16051CC','#5a9bd4CC']

# Border impact
ax= sns.violinplot(data=df, x="Critere_DP", y="DPratio", hue="Candidat", split=True, inner="quart", linewidth=1, palette=colors_list, legend = False, width=0.9)
sns.despine(left=True)
ax.set_xlabel("Border impact")
ax.set_ylabel("Counts")
ax.legend([],[], frameon=False)


# Plot visualisation
if not os.path.exists("images"):
    os.mkdir("images")
plt.savefig("images/PW_violin_DP_10samples.svg")
```

##### Graph C/ Filter candidates on candidat length.

Script to generate this graphic:

```
#!/usr/bin/env python3

import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import os

sns.set_theme(style="whitegrid")
df = pd.read_csv('data/cat_pw_candidates.Cglen.intron.Cdlen.DPratio.RI.JC.csv', sep=',')

# Colors list
colors_list=['#F16051CC','#5a9bd4']

# Select len impact
ax=sns.violinplot(data=df, x="Critere_lCd", y="lCd", hue="Candidat", split=True, inner="quart", linewidth=1, palette=colors_list, width=0.9)
sns.despine(left=True)
ax.set_xlabel("Candidat length")
ax.set_ylabel("Counts")
ax.legend([],[], frameon=False)

# Plot visualisation
if not os.path.exists("images"):
    os.mkdir("images")
plt.savefig("images/PW_violin_lCd_10samples.svg")
```

##### Graph D/ Filter candidates on retained intron ratio.

Retained intron ratio = (len contig / len candidat)

Script to generate this graphic:

```
#!/usr/bin/env python3

import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import os

sns.set_theme(style="whitegrid")
df = pd.read_csv('data/cat_pw_candidates.Cglen.intron.Cdlen.DPratio.RI.JC.csv', sep=',')

# Colors list
colors_list=['#F16051','#5a9bd4']

# Draw a nested violinplot and split the violins for easier comparison
ax=sns.violinplot(data=df, x="Critere_RI", y="RI", hue="Candidat", split=True, inner="quart", linewidth=1, palette=colors_list, legend = False)
sns.despine(left=True)
ax.set_xlabel("Candidat length/Contig length")
ax.set_ylabel("Counts")
ax.legend([],[], frameon=False)

# Plot visualisation
if not os.path.exists("images"):
    os.mkdir("images")
plt.savefig("images/PW_violin_RI_10samples.svg")
```


FRS_CAS-A/ Alignment statistics for STAR and HISAT2
----------------------------------------------------

##### 1 - Step 1 : Full Random Simulation

```diff
source activate ISeeker_environment

intronSeeker fullRandomSimulation -n 1000 -r  -o FRS_CAS-A/sample1/ -p sample1
intronSeeker fullRandomSimulation -n 1000 -r  -o FRS_CAS-A/sample2/ -p sample2
intronSeeker fullRandomSimulation -n 1000 -r  -o FRS_CAS-A/sample3/ -p sample3
intronSeeker fullRandomSimulation -n 1000 -r  -o FRS_CAS-A/sample4/ -p sample4
intronSeeker fullRandomSimulation -n 1000 -r  -o FRS_CAS-A/sample5/ -p sample5
```

Results:

```diff
(ISeeker_environment) $ ls FRS_CAS-A/sample2/
frs_sample2_contigs.fa  frs_sample2_contigs-modified.fa  frs_sample2_contigs-modified.gtf
```

##### 2 - Step 2 : Simulate Reads

```diff
intronSeeker simulateReads -o ./ -f frs_sample1_contigs.fa -c config/grinder_frs_testA.cfg


intronSeeker simulateReads -o ./ -f frs_sample2_contigs.fa -c config/grinder_frs_testA.cfg
intronSeeker simulateReads -o ./ -f frs_sample3_contigs.fa -c config/grinder_frs_testA.cfg
intronSeeker simulateReads -o ./ -f frs_sample4_contigs.fa -c config/grinder_frs_testA.cfg
intronSeeker simulateReads -o ./ -f frs_sample5_contigs.fa -c config/grinder_frs_testA.cfg
```

##### 3 - Step 3 : STAR and HISAT2 alignment and Split Read Search

```diff
(ISeeker_environment) $  intronSeeker starAlignment -r frs_sample1_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o STAR_alignment
(ISeeker_environment) $ ls 
frs_sample1_contigs.fa  frs_sample1_contigs-modified.fa  frs_sample1_contigs-modified.gtf  sr.log  sr_R1.fastq.gz  sr_R2.fastq.gz  sr_ranks.txt  STAR_alignment

(ISeeker_environment) $ intronSeeker splitReadSearch -a STAR_alignment/star.sort.bam -r frs_sample1_contigs-modified.fa -o sample1_splicing_event_STAR
(ISeeker_environment) $ ls
frs_sample1_contigs.fa           frs_sample1_contigs-modified.fa.fai  sample1_splicing_event_STAR  sr_R1.fastq.gz  sr_ranks.txt
frs_sample1_contigs-modified.fa  frs_sample1_contigs-modified.gtf     sr.log                       sr_R2.fastq.gz  STAR_alignment

(ISeeker_environment) $ intronSeeker hisat2Alignment -r frs_sample1_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o HISAT2_alignment
(ISeeker_environment) $ ls
frs_sample1_contigs.fa           frs_sample1_contigs-modified.fa.fai  HISAT2_alignment             sr.log          sr_R2.fastq.gz  STAR_alignment
frs_sample1_contigs-modified.fa  frs_sample1_contigs-modified.gtf     sample1_splicing_event_STAR  sr_R1.fastq.gz  sr_ranks.txt

(ISeeker_environment) $ intronSeeker splitReadSearch -a HISAT2_alignment/hisat2.sort.bam -r frs_sample1_contigs-modified.fa -o sample1_splicing_event_HISAT2
(ISeeker_environment) $ ls
frs_sample1_contigs.fa           frs_sample1_contigs-modified.fa.fai  HISAT2_alignment               sample1_splicing_event_STAR  sr_R1.fastq.gz  sr_ranks.txt
frs_sample1_contigs-modified.fa  frs_sample1_contigs-modified.gtf     sample1_splicing_event_HISAT2  sr.log                       sr_R2.fastq.gz  STAR_alignment


intronSeeker starAlignment -r frs_sample2_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o STAR_alignment
intronSeeker splitReadSearch -a STAR_alignment/star.sort.bam -r frs_sample2_contigs-modified.fa -o sample2_splicing_event_STAR
intronSeeker hisat2Alignment -r frs_sample2_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o HISAT2_alignment
intronSeeker splitReadSearch -a HISAT2_alignment/hisat2.sort.bam -r frs_sample2_contigs-modified.fa -o sample2_splicing_event_HISAT2
```

Same command lines for samples 3, 4 and 5.

##### IntronSeeker simulation Report 

For each sample:

```diff
python3 simulation2HTML.py -m /path/to/FRS_CAS-A/sample1/frs_sample1_contigs-modified.fa -f /path/to/FRS_CAS-A/sample1/frs_sample1_contigs.fa -g /path/to/CAS-A/sample1/frs_sample1_contigs-modified.gtf -o /path/to/FRS_CAS-A/sample1/HTML -p FRS_CASA_sample1_n1000_r_STAR -F  -1 /path/to/FRS_CAS-A/sample1/sr_R1.fastq.gz -2 /path/to/FRS_CAS-A/sample1/sr_R2.fastq.gz --flagstat /path/to/FRS_CAS-A/sample1/STAR_alignment/star.sort.flagstat.txt -c /path/to/FRS_CAS-A/sample1/sample1_splicing_event_STAR/srs_candidates.txt -r /path/to/FRS_CAS-A/sample1/sr_ranks.txt -s  /path/to/FRS_CAS-A/sample1/sample1_splicing_event_STAR/srs_split_alignments.txt -t 6
```

FRS_CAS-B/ Coverage impact
--------------------------

##### Test several coverages

For sample 1 (only), without -r fullRandomSimulation option (1 contig = 1 intron) nor -n, run several tests with coverage 3, 6, 9, 12, 15, 18, 21 :
Remarque : Without fullRandomSimulation -n option because we add -cf 3 in grinder config file.

```diff
$ more config/grinder_frs_testB-3.cfg
-cf 3                             
#-tr 2000000
-rd 101
-id 400 normal 100
-mo FR
-md poly4 3e-3 3.3e-8 
-mr 90 10
-am uniform
-fq 1
-ql 32 2
```

Warning :  « Coverage influence directly intron detection. In fact, the lower the coverage is, the harder the detection is. »

```diff
(ISeeker_environment) $ intronSeeker fullRandomSimulation  -o FRS_CAS-B/sample1/cf3/ -p sample1

(ISeeker_environment) $ intronSeeker simulateReads -o ./ -f frs_sample1_contigs.fa -c ../../config/grinder_frs_testB-3.cfg
(ISeeker_environment) $ intronSeeker starAlignment -r frs_sample1_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o STAR_alignment
(ISeeker_environment) $ intronSeeker splitReadSearch -a STAR_alignment/star.sort.bam -r frs_sample1_contigs-modified.fa -o sample1_splicing_event_STAR
(ISeeker_environment) $ intronSeeker hisat2Alignment -r frs_sample1_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o HISAT2_alignment
(ISeeker_environment) $ intronSeeker splitReadSearch -a HISAT2_alignment/hisat2.sort.bam -r frs_sample1_contigs-modified.fa -o sample1_splicing_event_HISAT2
```

Same command lines for cf 6, 9, 12, 15, 18 and 21.


##### IntronSeeker simulation Report 

For each coverage:

```diff
python3 simulation2HTML.py -m /path/to/FRS_CAS-B/sample1/cf3/frs_sample1_contigs-modified.fa -f /path/to/FRS_CAS-B/sample1/cf3/frs_sample1_contigs.fa -g /path/to/FRS_CAS-B/sample1/cf3/frs_sample1_contigs-modified.gtf -o /path/to/FRS_CAS-B/sample1/cf3/HTML -p FRS_CASB_sample1_cf3 -F  -1 /path/to/FRS_CAS-B/sample1/cf3/sr_R1.fastq.gz -2 /path/to/FRS_CAS-B/sample1/cf3/sr_R2.fastq.gz --flagstat /path/to/FRS_CAS-B/sample1/cf3/STAR_alignment/star.sort.flagstat.txt -c /path/to/FRS_CAS-B/sample1/cf3/sample1_splicing_event_STAR/srs_candidates.txt -r /path/to/FRS_CAS-B/sample1/cf3/sr_ranks.txt -s  /path/to/FRS_CAS-B/sample1/cf3/sample1_splicing_event_STAR/srs_split_alignments.txt -t 6
```

FRS_CAS-C/ Relative Abundance impact
------------------------------------

Objective : Count number of input reference sequences by expression class with deux differents abundance models : uniform and powerlaw.

##### Test with relative abundance model uniform

```diff
FRS_CAS-C/config/grinder_frs_testC-am-uniform.cfg
-tr 2000000
-rd 100 normal 7.5
-id 500 normal 250
-mo FR
-md poly4 3e-3 3.3e-8 
-mr 80 20 
-am uniform
-fq 1
-ql 32 2

(ISeeker_environment) $ intronSeeker fullRandomSimulation -n 1000 -r -o FRS_CAS-C/sample1/am/ -p sample1

(ISeeker_environment) $ intronSeeker simulateReads -o ./ -f frs_sample1_contigs.fa -c ../../config/grinder_frs_testC-am-uniform.cfg
(ISeeker_environment) $ intronSeeker starAlignment -r frs_sample1_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o STAR_alignment
(ISeeker_environment) $ intronSeeker splitReadSearch -a STAR_alignment/star.sort.bam -r frs_sample1_contigs-modified.fa -o sample1_splicing_event_STAR
(ISeeker_environment) $ intronSeeker hisat2Alignment -r frs_sample1_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o HISAT2_alignment
(ISeeker_environment) $ intronSeeker splitReadSearch -a HISAT2_alignment/hisat2.sort.bam -r frs_sample1_contigs-modified.fa -o sample1_splicing_event_HISAT2
```

##### IntronSeeker simulation Report 

```diff
python3 simulation2HTML.py -m /path/to/FRS_CAS-C/sample1/am/frs_sample1_contigs-modified.fa -f /path/to/FRS_CAS-C/sample1/am/frs_sample1_contigs.fa -g /path/to/FRS_CAS-C/sample1/am/frs_sample1_contigs-modified.gtf -o /path/to/FRS_CAS-C/sample1/am/HTML -p FRS_CASC_sample1_am -F  -1 /path/to/FRS_CAS-C/sample1/am/sr_R1.fastq.gz -2 /path/to/FRS_CAS-C/sample1/am/sr_R2.fastq.gz --flagstat /path/to/FRS_CAS-C/sample1/am/STAR_alignment/star.sort.flagstat.txt -c /path/to/FRS_CAS-C/sample1/am/sample1_splicing_event_STAR/srs_candidates.txt -r /path/to/FRS_CAS-C/sample1/am/sr_ranks.txt -s  /path/to/FRS_CAS-C/sample1/am/sample1_splicing_event_STAR/srs_split_alignments.txt -t 6
```

##### Test with relative abundance model powerlaw 0.5

```diff
FRS_CAS-C/config/grinder_frs_testC-pw05.cfg
-tr 2000000
-rd 100 normal 7.5
-id 500 normal 250
-mo FR
-md poly4 3e-3 3.3e-8 
-mr 80 20 
-am powerlaw 0.5
-fq 1
-ql 32 2

(ISeeker_environment) $ intronSeeker fullRandomSimulation -n 1000 -r -o FRS_CAS-C/sample1/pw05/ -p sample1

(ISeeker_environment) $ intronSeeker simulateReads -o ./ -f frs_sample1_contigs.fa -c ../../config/grinder_frs_testC-pw05.cfg
(ISeeker_environment) $ intronSeeker starAlignment -r frs_sample1_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o STAR_alignment
(ISeeker_environment) $ intronSeeker splitReadSearch -a STAR_alignment/star.sort.bam -r frs_sample1_contigs-modified.fa -o sample1_splicing_event_STAR
(ISeeker_environment) $ intronSeeker hisat2Alignment -r frs_sample1_contigs-modified.fa -1 sr_R1.fastq.gz -2 sr_R2.fastq.gz -o HISAT2_alignment
(ISeeker_environment) $ intronSeeker splitReadSearch -a HISAT2_alignment/hisat2.sort.bam -r frs_sample1_contigs-modified.fa -o sample1_splicing_event_HISAT2
```

##### IntronSeeker simulation Report 

```diff
python3 simulation2HTML.py -m /path/to/FRS_CAS-C/sample1/pw05/frs_sample1_contigs-modified.fa -f /path/to/FRS_CAS-C/sample1/pw05/frs_sample1_contigs.fa -g /path/to/FRS_CAS-C/sample1/pw05/frs_sample1_contigs-modified.gtf -o /path/to/FRS_CAS-C/sample1/pw05/HTML -p FRS_CASC_sample1_pw05 -F  -1 /path/to/FRS_CAS-C/sample1/pw05/sr_R1.fastq.gz -2 /path/to/FRS_CAS-C/sample1/pw05/sr_R2.fastq.gz --flagstat /path/to/FRS_CAS-C/sample1/pw05/STAR_alignment/star.sort.flagstat.txt -c /path/to/FRS_CAS-C/sample1/pw05/sample1_splicing_event_STAR/srs_candidates.txt -r /path/to/FRS_CAS-C/sample1/pw05/sr_ranks.txt -s  /path/to/FRS_CAS-C/sample1/pw05/sample1_splicing_event_STAR/srs_split_alignments.txt -t 6
```